### ISSUE TITLE
PDFDecrypt: "AttributeError: 'str' object has no attribute 'decode'"

### ISSUE THREAD

**Issue by Geremia** [apprenticeharper/DeDRM_tools#1272]
When decrypting an ADE PDF, I get this:
```
DeDRM v7.0.0: Trying to decrypt ADE_PDF.pdf
DeDRM v7.0.0: ADE_PDF.pdf is a PDF ebook
Running file type plugin DeDRM failed with traceback:
Traceback (most recent call last):
  File "site-packages/calibre/customize/ui.py", line 172, in _run_filetype_plugins
  File "calibre_plugins.dedrm.__init__", line 633, in run
  File "calibre_plugins.dedrm.__init__", line 423, in PDFDecrypt
AttributeError: 'str' object has no attribute 'decode'
```
How do I locate in the uncompiled Python code where this decode() is being called on a string?
I was able to successfully decrypt PDFs and EPUBs with a pre-5.0 Calibre version and DeDRM 6.8.0. I'm using ADE 4.5.126523, and I changed no keys.

**Comment by ElleKayEm** [comment]
Line 423 here:

https://github.com/apprenticeharper/DeDRM_tools/blob/master/DeDRM_plugin/__init__.py

**Comment by Geremia** [comment]
@ElleKayEm 

I changed that line to:
``` python3
userkey = bytearray.fromhex(userkeyhex)
```

Now I get:
```
DeDRM v7.0.0: Trying to decrypt ADE_PDF.pdf
DeDRM v7.0.0: ADE_PDF.pdf is a PDF ebook
DeDRM v7.0.0: Trying Encryption key adobekey_1
DeDRM v7.0.0: Exception when decrypting after 0.0 seconds
Traceback (most recent call last):
  File "calibre_plugins.dedrm.__init__", line 429, in PDFDecrypt
  File "calibre_plugins.dedrm.ineptpdf", line 2182, in decryptBook
  File "calibre_plugins.dedrm.ineptpdf", line 1987, in __init__
  File "calibre_plugins.dedrm.ineptpdf", line 1795, in __init__
  File "calibre_plugins.dedrm.ineptpdf", line 1420, in set_parser
  File "calibre_plugins.dedrm.ineptpdf", line 1921, in read_xref
  File "calibre_plugins.dedrm.ineptpdf", line 1872, in find_xref
  File "calibre_plugins.dedrm.ineptpdf", line 858, in revreadlines
TypeError: argument should be integer or bytes-like object, not 'str'
DeDRM v7.0.0: Failed to decrypt with key adobekey_1 after 0.0 seconds
DeDRM v7.0.0: Trying Encryption key default_key
DeDRM v7.0.0: Exception when decrypting after 0.0 seconds
```

**Comment by tartley** [comment]
@Geremia When that error is fixed, there will likely be a bunch more such problems. The whole body of code is gradually being converted from Python2 to Python3.

It's a bit of an unfortunate situation: It's possible for you to fix each error as you see it, converting the code to Python3 in the process, but that runs the risk of duplicating work that someone else is already doing. In particular, I understand the esteemed @apprenticeharper has a significant branch in progress. I started a branch of my own, to fix the scenario I was interested in, but very understandably, it was closed when it turned out to be largely duplicating other work. Under different circumstances one could imagine it being easier for motivated individuals to help out, but that seems difficult to organize at this stage.

Sticking with Calibre v4.x is the standard advice, until this difficult transition has passed.
